src/main/java/server/ChatServer.java

```diff
@@ -142,82 +99,113 @@ public static void onlineMessage() throws IOException {
     Socket s;
     boolean isloggedin;
 
+
+    /**
+     * Try to accept the connection. Returns false if the connection is rejected. for any reason.
+     * @param userService
+     * @return
+     */
+    static ClientHandler tryAccept(Socket s, UserService userService) {
+        DataInputStream dis = new DataInputStream(s.getInputStream());
+        PrintWriter dos = new PrintWriter(s.getOutputStream(),true);
+        StringTokenizer st = new StringTokenizer(dis.readLine(), "#");
+        String cmd = st.nextToken();
+
+        try {
+            if (cmd.equals("CONNECT")) {
+                String username = st.nextToken();
+                if (!userService.usernameExists(username)) {
+                    dos.println("CLOSE#2");
+                    s.close();
+                    break;
+                }
+
+                return new ClientHandler(s, usenrame, dis, dos);
+            }
+
+            dos.println("CLOSE#1");
+        } catch (Exception e) {
+            // Intentionally swallowing the exception.
+        } finally {
+            // Hitting here means that the client failed to be accepted.
+            return null;
+        }
+    }
+
     // constructor
     public ClientHandler(Socket s, String name, DataInputStream dis, PrintWriter dos) {
         this.dis = dis;
         this.dos = dos;
         this.name = name;
         this.s = s;
         this.isloggedin = true;
-
     }
 
     public String getName() {
         return name;
     }
 
-    public void close(int i) throws IOException {
+    // nameMatches indicates whether or not the given pattern matches the ClientHandlers name.
+    public boolean nameMatches(String pattern) {
```

Comment:
We handle name matching here. Note that we put the two simple cases first and the more expensive one last.